### PR TITLE
Add integration test cases for backup and backing chain

### PR DESCRIPTION
**Pull request by lento-sun**
To test incremental backup with backing chain management functions:
blockpull, blockcommit and blockcopy.

Signed-off-by: Yi Sun <yisun@redhat.com>

### CODE REVIEW

**Comment by lento-sun** [comment]
depend on following PRs:
avocado-framework/avocado-vt#2503
avocado-framework/avocado-vt#2506
avocado-framework/avocado-vt#2507

**Comment by lento-sun** [comment]
test result as follow:
```
(.libvirt-ci-venv-ci-runtest-tISC7i) [root@hp-dl320eg8-10 virttest]# avocado run --vt-type libvirt type_specific.io-github-autotest-libvirt.incremental_backup.backing_chain
JOB ID     : 78fd820539e20e3b8385902b9e27b34e914377ba
JOB LOG    : /root/avocado/job-results/job-2020-06-04T08.41-78fd820/job.log
 (01/12) type_specific.io-github-autotest-libvirt.incremental_backup.backing_chain.blockpull.mid_to_top.original_disk_local.hotplug_disk.nbd_tcp.scratch_to_file: PASS (40.20 s)
 (02/12) type_specific.io-github-autotest-libvirt.incremental_backup.backing_chain.blockpull.mid_to_top.original_disk_local.coldplug_disk.nbd_tcp.scratch_to_file: PASS (39.52 s)
 (03/12) type_specific.io-github-autotest-libvirt.incremental_backup.backing_chain.blockpull.base_to_top.original_disk_local.hotplug_disk.nbd_tcp.scratch_to_file: ERROR: Command '/bin/virsh checkpoint-delete avocado-vt-vm1 checkpoint_0 ' failed.\nstdout: b'\n'\nstderr: b"error: Failed to delete checkpoint checkpoint_0\nerror: internal error: bitmap 'checkpoint_0' not found in backing chain of 'vdb'\n"\nadditional_info: None (38.77 s)
 (04/12) type_specific.io-github-autotest-libvirt.incremental_backup.backing_chain.blockpull.base_to_top.original_disk_local.coldplug_disk.nbd_tcp.scratch_to_file: ERROR: Command '/bin/virsh checkpoint-delete avocado-vt-vm1 checkpoint_0 ' failed.\nstdout: b'\n'\nstderr: b"error: Failed to delete checkpoint checkpoint_0\nerror: internal error: bitmap 'checkpoint_0' not found in backing chain of 'vdb'\n"\nadditional_info: None (39.70 s)
 (05/12) type_specific.io-github-autotest-libvirt.incremental_backup.backing_chain.blockcommit.top_to_base.original_disk_local.hotplug_disk.nbd_tcp.scratch_to_file: PASS (40.68 s)
 (06/12) type_specific.io-github-autotest-libvirt.incremental_backup.backing_chain.blockcommit.top_to_base.original_disk_local.coldplug_disk.nbd_tcp.scratch_to_file: PASS (40.91 s)
 (07/12) type_specific.io-github-autotest-libvirt.incremental_backup.backing_chain.blockcommit.mid_to_base.original_disk_local.hotplug_disk.nbd_tcp.scratch_to_file: PASS (40.86 s)
 (08/12) type_specific.io-github-autotest-libvirt.incremental_backup.backing_chain.blockcommit.mid_to_base.original_disk_local.coldplug_disk.nbd_tcp.scratch_to_file: PASS (40.31 s)
 (09/12) type_specific.io-github-autotest-libvirt.incremental_backup.backing_chain.blockcommit.mid_to_mid.original_disk_local.hotplug_disk.nbd_tcp.scratch_to_file: PASS (43.81 s)
 (10/12) type_specific.io-github-autotest-libvirt.incremental_backup.backing_chain.blockcommit.mid_to_mid.original_disk_local.coldplug_disk.nbd_tcp.scratch_to_file: PASS (44.09 s)
 (11/12) type_specific.io-github-autotest-libvirt.incremental_backup.backing_chain.blockcommit.top_to_mid.original_disk_local.hotplug_disk.nbd_tcp.scratch_to_file: PASS (40.76 s)
 (12/12) type_specific.io-github-autotest-libvirt.incremental_backup.backing_chain.blockcommit.top_to_mid.original_disk_local.coldplug_disk.nbd_tcp.scratch_to_file: PASS (41.03 s)
RESULTS    : PASS 10 | ERROR 2 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 494.03 s
```
Failed cases are due to existing bz: https://bugzilla.redhat.com/show_bug.cgi?id=1799010 (incremental-backup: RFE: Handle backup bitmaps during virDomainBlockPull)

**Comment by lento-sun** [comment]
more cases added, test result can be found at: https://libvirt-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/view/libvirt/view/RHEL-8.3%20x86_64/job/libvirt-RHEL-8.3-runtest-x86_64-function-incremental_backup/10/testReport/rhel/incremental_backup/
with pattern: incremental_backup.backing_chain
Added cases to test externally created snapshot:
```
incremental_backup.backing_chain.blockpull.mid_to_top.original_disk_local.nbd_tcp.shutoff_snapshot.scratch_to_file
incremental_backup.backing_chain.blockpull.base_to_top.original_disk_local.nbd_tcp.shutoff_snapshot.scratch_to_file
incremental_backup.backing_chain.blockcommit.top_to_base.original_disk_local.nbd_tcp.shutoff_snapshot.scratch_to_file
incremental_backup.backing_chain.blockcommit.mid_to_base.original_disk_local.nbd_tcp.shutoff_snapshot.scratch_to_file
incremental_backup.backing_chain.blockcommit.mid_to_mid.original_disk_local.nbd_tcp.shutoff_snapshot.scratch_to_file
incremental_backup.backing_chain.blockcommit.top_to_mid.original_disk_local.nbd_tcp.shutoff_snapshot.scratch_to_file
incremental_backup.backing_chain.blockcopy.reuse_external.full_copy.original_disk_local.nbd_tcp.shutoff_snapshot.scratch_to_file
incremental_backup.backing_chain.blockcopy.reuse_external.shallow_copy.original_disk_local.nbd_tcp.shutoff_snapshot.scratch_to_file
incremental_backup.backing_chain.blockcopy.not_reuse_external.full_copy.original_disk_local.nbd_tcp.shutoff_snapshot.scratch_to_file
incremental_backup.backing_chain.blockcopy.not_reuse_external.shallow_copy.original_disk_local.nbd_tcp.shutoff_snapshot.scratch_to_file
```

**Comment by dzhengfy** [comment]
@lento-sun Please help provide some description for this PR.

File: libvirt/tests/cfg/incremental_backup/incremental_backup_backing_chain.cfg [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    variants:
+        - original_disk_local:
+            original_disk_type = "local"
+    variants blockcommand:
+        - blockpull:
+            variants from_to:
+                - mid_to_top:
+                - base_to_top:
+        - blockcommit:
+            variants from_to:
```
Comment (chunfuwen) [inline libvirt/tests/cfg/incremental_backup/incremental_backup_backing_chain.cfg:27]: will variant from_to in this line override from_to variant in line 33 ?

File: libvirt/tests/src/incremental_backup/incremental_backup_backing_chain.py [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+        if blockcommand == "blockpull":
+            if from_to == "mid_to_top":
+                cmd_option = ("--base {0}[{1}] --wait").format(original_disk_target,
+                                                               middle_layer1_index)
+            elif from_to == "base_to_top":
+                cmd_option = ("--base {0}[{1}] --wait").format(original_disk_target,
+                                                               base_layer_index)
+            virsh.blockpull(vm_name, original_disk_target, cmd_option,
+                            debug=True, ignore_status=False)
+        if blockcommand == "blockcommit":
```
Comment (chunfuwen) [inline libvirt/tests/src/incremental_backup/incremental_backup_backing_chain.py:73]: will you consider wrap each blockcommand part logic  as one method?

File: libvirt/tests/src/incremental_backup/incremental_backup_backing_chain.py [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+                    create_img_cmd = "qemu-img create -f qcow2 -F qcow2 -b %s %s"
+                    create_img_cmd %= (backend_img, copy_dest)
+                else:
+                    create_img_cmd = "qemu-img create -f qcow2 %s %s"
+                    create_img_cmd %= (copy_dest, original_disk_size)
+                process.run(create_img_cmd, shell=True, ignore_status=False)
+            virsh.blockcopy(vm_name, original_disk_target, copy_dest, cmd_option,
+                            debug=True, ignore_status=False)
+
+    def create_shutoff_snapshot(original_img, snapshot_img):
```
Comment (chunfuwen) [inline libvirt/tests/src/incremental_backup/incremental_backup_backing_chain.py:124]: miss python doc string

File: libvirt/tests/src/incremental_backup/incremental_backup_backing_chain.py [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+        except Exception as bitmap_error:
+            logging.debug("Cannot add bitmap to new image, skip it: %s",
+                          bitmap_error)
+
+    # Cancel the test if libvirt version is too low
+    if not libvirt_version.version_compare(6, 0, 0):
+        test.cancel("Current libvirt version doesn't support "
+                    "incremental backup.")
+
+    original_disk_size = params.get("original_disk_size", "100M")
```
Comment (chunfuwen) [inline libvirt/tests/src/incremental_backup/incremental_backup_backing_chain.py:147]: Better orgornize those below paremater into a few groups and add one comment before each group.

**Review by chunfuwen — COMMENTED** [review]
see inline comment

File: libvirt/tests/cfg/incremental_backup/incremental_backup_backing_chain.cfg (hunk 1 above)
Comment (lento-sun) [reply libvirt/tests/cfg/incremental_backup/incremental_backup_backing_chain.cfg:27]: do you mean line 23?
blockpull and blockcommit will never overlap in this script, so the from_to will not be overridden

Comment (lento-sun) [reply libvirt/tests/cfg/incremental_backup/incremental_backup_backing_chain.cfg:27]: do you mean line 23?
blockpull and blockcommit will never overlap in this script, so the from_to will not be overridden

File: libvirt/tests/src/incremental_backup/incremental_backup_backing_chain.py (hunk 2 above)
Comment (lento-sun) [reply libvirt/tests/src/incremental_backup/incremental_backup_backing_chain.py:73]: Updated to run_blk_cmd()'s local methods

File: libvirt/tests/src/incremental_backup/incremental_backup_backing_chain.py (hunk 3 above)
Comment (lento-sun) [reply libvirt/tests/src/incremental_backup/incremental_backup_backing_chain.py:124]: added

File: libvirt/tests/src/incremental_backup/incremental_backup_backing_chain.py (hunk 4 above)
Comment (lento-sun) [reply libvirt/tests/src/incremental_backup/incremental_backup_backing_chain.py:147]: updated

**Comment by lento-sun** [comment]
> @lento-sun Please help provide some description for this PR.

Done

**Review by chunfuwen — APPROVED** [review]
LGTM

**Review by dzhengfy — APPROVED** [review]
LGTM

**Pull request merged**